### PR TITLE
core: Fix confusing behaviors of FROM_BASE64 function in WHERE clause

### LINKED ISSUE #28154: Confusing behaviors of FROM_BASE64 function in WHERE clause
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
/* init */ drop table if exists t;
/* init */ create table t(a TEXT);
/* init */ insert into t values('abc');

tidb> select * from t where from_base64('');  -- 1
tidb> update t set a = 'def' where from_base64('');  -- 2
tidb> select * from t where from_base64('invalidbase64');  -- 3
tidb> update t set a = 'hig' where from_base64('invalidbase64');  -- 4
tidb> select * from t where from_base64('test');  -- 5
tidb> update t set a = 'xyz' where from_base64('test');  -- 6
```
It should be noted that 'test' is a valid base64 string:
```
tidb> select from_base64('test');
+------------------------------------------+
| from_base64('test')                      |
+------------------------------------------+
| 0xB5EB2D                                 |
+------------------------------------------+
1 row in set (0.00 sec)
```
### 2. What did you expect to see? (Required)

If the argument of `FROM_BASE64()` is a invalid base64 string, the function should be evaluated as `NULL`, so the result of statement 1, 2, 3, 4 should be empty.

If the argument of `FROM_BASE64()` is a valid base64 string, the function's result should be `0xB5EB2D`, so the  statement 4, 5 should cause warning and error respectively.

### 3. What did you see instead (Required)
```
tidb> select * from t where from_base64('');  -- 1
Empty set, 1 warning (0.00 sec)

tidb> update t set a = 'def' where from_base64('');  -- 2
Query OK, 1 row affected (0.06 sec)
Rows matched: 1  Changed: 1  Warnings: 0

tidb> select * from t where from_base64('invalidbase64');  -- 3
Empty set (0.00 sec)

tidb> update t set a = 'def' where from_base64('invalidbase64');  -- 4
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

tidb> select * from t where from_base64('test');  -- 5
Empty set, 1 warning (0.01 sec)

tidb> update t set a = 'xyz' where from_base64('test');  -- 6
Query OK, 1 row affected (0.04 sec)
Rows matched: 1  Changed: 1  Warnings: 0
```

As you can see, `from_base64()` with a empty string was evaluated as `False`(with a warning) in SELECT but was evaluated as `True` in UPDATE.  Besides, the results of statement 5 and 6 means that a valid base64 string in `from_base64()` will be evaluated as `False`(with a warning) in SELECT but will be evaluated as `True` in UPDATE. I think this inconsistent behaviors can cause terrible consequences. 

### 4. What is your TiDB version? (Required)
```
mysql> select version();
+--------------------+
| version()          |
+--------------------+
| 5.7.25-TiDB-v5.2.1 |
+--------------------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by zhaoxugang**
1. core: Fix confusing behaviors of FROM_BASE64 function in WHERE clause

### What problem does this PR solve?

Issue Number: close #28154 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Coorercing the function `FROM_BASE64()` to a boolean value could return invalid results.
```

### CODE REVIEW

**Comment by zhaoxugang** [comment]
> ```sql
> select * from t;
> ```
I tried in mysql 8.0.21,following:
mysql> insert into t values('abc');
Query OK, 1 row affected (0.00 sec)

mysql> select * from t where from_base64('');
Empty set (0.00 sec)

mysql> update t set a = 'def' where from_base64('');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

mysql> select * from t where from_base64('invalidbase64');
Empty set (0.00 sec)

mysql> update t set a = 'hig' where from_base64('invalidbase64');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

mysql> select * from t where from_base64('test');
Empty set (0.00 sec)

mysql> update t set a = 'xyz' where from_base64('test');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0  Changed: 0  Warnings: 0

mysql> select * from t;
+------+
| a    |
+------+
| abc  |
+------+
1 row in set (0.01 sec)

mysql> select @@version;
+-----------+
| @@version |
+-----------+
| 8.0.21    |
+-----------+
1 row in set (0.00 sec)

**Comment by zhaoxugang** [comment]
> MySQL 5.7 and MySQL 8.0 behave differently here:
> 
> ```sql
> mysql [localhost:5731] {msandbox} (test) > select * from t where from_base64('');
> Empty set (0.00 sec)
> 
> mysql [localhost:5731] {msandbox} (test) > update t set a = 'def' where from_base64('');
> Query OK, 0 rows affected (0.00 sec)
> Rows matched: 0  Changed: 0  Warnings: 0
> 
> mysql [localhost:5731] {msandbox} (test) > select * from t where from_base64('invalidbase64');
> Empty set (0.00 sec)
> 
> mysql [localhost:5731] {msandbox} (test) > update t set a = 'hig' where from_base64('invalidbase64');
> Query OK, 0 rows affected (0.00 sec)
> Rows matched: 0  Changed: 0  Warnings: 0
> 
> mysql [localhost:5731] {msandbox} (test) > select * from t where from_base64('test');
> Empty set (0.00 sec)
> 
> mysql [localhost:5731] {msandbox} (test) > update t set a = 'xyz' where from_base64('test');
> Query OK, 0 rows affected (0.00 sec)
> Rows matched: 0  Changed: 0  Warnings: 0
> 
> mysql [localhost:5731] {msandbox} (test) > select * from t;
> +------+
> | a    |
> +------+
> | abc  |
> +------+
> 1 row in set (0.01 sec)
> ```
> 
> Versus:
> 
> ```sql
> mysql [localhost:8024] {root} (test) > insert into t values('abc');
> Query OK, 1 row affected (0.00 sec)
> 
> mysql [localhost:8024] {root} (test) > select * from t where from_base64('');
> Empty set (0.00 sec)
> 
> mysql [localhost:8024] {root} (test) > update t set a = 'def' where from_base64('');
> Query OK, 0 rows affected (0.00 sec)
> Rows matched: 0  Changed: 0  Warnings: 0
> 
> mysql [localhost:8024] {root} (test) > select * from t where from_base64('invalidbase64');
> Empty set (0.00 sec)
> 
> mysql [localhost:8024] {root} (test) > update t set a = 'hig' where from_base64('invalidbase64');
> Query OK, 0 rows affected (0.00 sec)
> Rows matched: 0  Changed: 0  Warnings: 0
> 
> mysql [localhost:8024] {root} (test) > select * from t where from_base64('test');
> Empty set, 1 warning (0.00 sec) <-- the warning is a truncated value
> 
> mysql [localhost:8024] {root} (test) > update t set a = 'xyz' where from_base64('test');
> ERROR 1292 (22007): Truncated incorrect DOUBLE value: '\xB5\xEB-'
> mysql [localhost:8024] {root} (test) > select * from t;
> +------+
> | a    |
> +------+
> | abc  |
> +------+
> 1 row in set (0.00 sec)
> ```

I tried in mysql 8.0.21,following:
mysql> insert into t values('abc');
Query OK, 1 row affected (0.00 sec)

mysql> select * from t where from_base64('');
Empty set (0.00 sec)

mysql> update t set a = 'def' where from_base64('');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0 Changed: 0 Warnings: 0

mysql> select * from t where from_base64('invalidbase64');
Empty set (0.00 sec)

mysql> update t set a = 'hig' where from_base64('invalidbase64');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0 Changed: 0 Warnings: 0

mysql> select * from t where from_base64('test');
Empty set (0.00 sec)

mysql> update t set a = 'xyz' where from_base64('test');
Query OK, 0 rows affected (0.00 sec)
Rows matched: 0 Changed: 0 Warnings: 0

mysql> select * from t;
+------+
| a |
+------+
| abc |
+------+
1 row in set (0.01 sec)

mysql> select @@Version;
+-----------+
| @@Version |
+-----------+
| 8.0.21 |
+-----------+
1 row in set (0.00 sec)

**Review by morgo — APPROVED** [review]
LGTM, I've updated the release note

**Review by xhebox — APPROVED** [review]
LGTM

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -987,7 +987,10 @@ func (b *PlanBuilder) buildSelection(ctx context.Context, p LogicalPlan, where a
 		for _, item := range cnfItems {
 			if con, ok := item.(*expression.Constant); ok && con.DeferredExpr == nil && con.ParamMarker == nil {
 				ret, _, err := expression.EvalBool(b.ctx, expression.CNFExprs{con}, chunk.Row{})
-				if err != nil || ret {
+				if err != nil {
+					return nil, errors.Trace(err)
```
Comment (xiongjiwei) [inline planner/core/logical_plan_builder.go:991]: return error here may cause problems
```sql
update t set a = 'b' where ' ';
```
in tidb, it returns error, in MySQL it is ok.

Comment (xiongjiwei) [reply planner/core/logical_plan_builder.go:991]: @zhaoxugang PTAL

Comment (yuqi1129) [reply planner/core/logical_plan_builder.go:991]: +1

Comment (zhaoxugang) [reply planner/core/logical_plan_builder.go:991]: IMO the main question is that when convert err is present we should return err or return empty set？if return err maybe effect the user of tidb if return empty will confuse with actual return empty set.I need some advices
@xiongjiwei @yuqi1129 @morgo

**Pull request merged**